### ISSUE TITLE
Contexts and merging of multi paths

### ISSUE THREAD

**Issue by reyerm** [Botmock/botmock-dialogflow-export#46]
**Issue Overview**
Input context is incorrect in some cases where paths merge.
Additionally some intents are left with no Quick Reply or similar option for user to select to continue the conversation flow.

**To Reproduce**
See notes in email sent directly.

**Expected behavior**
Dialogflow is able to understand and navigate the merging of divergent paths back to common pathway.

**Screenshots**
See email

**Desktop (please complete the following information):**
 - NPM Version:

**Additional Context**

**Comment by reyerm** [comment]
out of interest, why linking to #42, as that had to do with support for images?

**Comment by nonnontrivial** [comment]
Sure @reyerm . I've been unable to reproduce the issue with images; the script did not produce errors and the zip imported into Dialogflow. At the moment, it feels much more likely that the problem you encountered with images should be related to this issue in some way.

Are you able to recreate the issue with images in a simple project?

**Comment by reyerm** [comment]
I’ll relinked images and give it a try now and let you know.

**Comment by reyerm** [comment]
Yes, it seems to be working fine now. Import works and when executed on small test in Dialogflow that works too. So hopefully image issue is now fully resolved.

Context issue remains.

**Comment by reyerm** [comment]
![image](https://user-images.githubusercontent.com/52199251/63355547-97100200-c366-11e9-8dac-5f03705a9af0.png)
As another example, a case where divergent paths rejoin, but the resultin quickreplies are only mapped in one of the intents (the tick) while the other path doesnt have the quick replies repeated in it and so the user cannot continue.

**Comment by reyerm** [comment]
Kevin: I have been reflecting on this challenge. My take is that to get correct mapping in Dialogflow, the algorithm needs to follow the logic of:
1. for every next intent being mapped, follow all the child objects until you hit another intent. all text response until that point to be included in the intent being mapped as responses and finally the quick reply or similar response to be the last item included in the intent being mapped. so recursively trace down the tree for all branches until we hit an intent on each branch. so we have a number of intents derived from this for each branch, and they would likely (intentionally) include duplicated responses and each one would hold the final quick reply or utterance. 
note: if we dont end up with duplication (same response/s on various intents) then we will hit deadends where one path ends without any final response/question that the user can respond to.
2. for the context, we reverse the trace back up all of the paths of tree that lead here, until we hit the previous intent(s) and that (each one we end at) becomes one of the input context. we can remove duplicate contexts if they arise, as only needed once in input and output context fields in dialogflow.

Does that make sense, and is that what is mean to be happening? or if not, would that solve out issue? happy to chat through by call.

note: an alternative or quick fix may be some design rules in Botmock that require you to include intents as last step before every path merges to ensure it can be mapped correcly, but that may lead to a cumbersome design with unnecessary intents/quick replies/etc.

**Comment by reyerm** [comment]
I wonder if the unnecessary (older contexts) could be contributing to my issues? as an example:
![image](https://user-images.githubusercontent.com/52199251/63500382-0ce0ae80-c4ca-11e9-8b72-f15d7c8e5de9.png)

As it Dialogflow doesnt understand the response. If however i edit the intent and remove the older input context qr_change_area then it works (somewhat surprisingly).

**Comment by reyerm** [comment]
Maybe easier to hop on a quick call with you and Kevin to explain/illustrate? 
Simple answer is I have had to fix some of the intent contexts in dialogflow as well as some responses, to get it to work in dialogflow. Seems some unusual situations that probably don’t come up in your current test cases?

**Comment by reyerm** [comment]
Ps to your question 2 and 3:
2 - no that isn’t what I mean. What I mean is some paths never get there. And some intents never fire.  
3 - there are a number of other cases I have hit but have just fixed them and not categorized / checked if similar scenarios to the ones I posted. 


Kind regards,
Reyer

**Comment by nonnontrivial** [comment]
Hi @reyerm . This repo has undergone a somewhat major refactor in the last few days. Can you `git pull`, then `npm install` and confirm that you still have this issue?

**Comment by reyerm** [comment]
Yes thanks, have been watching all the changes coming through with interest. Glad to hear it is ready to test again, I will hopefully get a chance to later this week.

To clarify, has some specific fix gone in to try rectify the incorrect “contexts” for multipath cases?
